File: src/main/java/com/acme/GraalPyService.java
package com.acme;

import org.graalvm.polyglot.Context;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class GraalPyService {
    private final GraalPyContextProvider provider;
    private Context pythonCtx;

    @Autowired
    public GraalPyService(GraalPyContextProvider provider) {
        this.provider = provider;
    }

    private Context getPythonCtx() {
        if (pythonCtx == null) {
            pythonCtx = provider.pythonContext();
            pythonCtx.eval("python", "from python_functions import *");
        }
        return pythonCtx;
    }

    public String greet(String name) {
        return getPythonCtx().getBindings("python")
            .getMember("greet")
            .execute(name)
            .asString();
    }
}

File: src/main/kotlin/com/acme/GraalPyService.kt
package com.acme

import io.smallrye.common.annotation.Blocking
import io.quarkus.runtime.Startup
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import org.graalvm.polyglot.Context
import org.graalvm.polyglot.Value

@Blocking
@Startup
@ApplicationScoped
class GraalPyService @Inject constructor(
    private val provider: GraalPyContextProvider
) {
    private val pythonCtx: Context by lazy {
        provider.pythonContext().apply {
            eval("python", "from python_functions import *")
        }
    }

    fun greet(name: String): String =
        pythonCtx.getBindings("python").getMember("greet").execute(name).asString()
}
